mwtext/wikitext_preprocessor_mwpfh.py

```diff
@@ -0,0 +1,290 @@
+"""
+Parse the wikitext representation of a Wikipedia page into paragraph objects.
+
+The goal of this module is to parse wikitext into plaintext in a way that preserves
+information about link offsets and section information.  The default settings are
+geared towards extracting the offsets for every internal link (i.e. links that lead
+to other main space pages in the same wiki).
+
+The actual parsing relies on the MediaWiki Parser From Hell.  Construction of the
+paragraph objects is heavily inspired by the approach taken in Wikipedia2Vec
+
+ * https://github.com/earwig/mwparserfromhell
+ * https://github.com/wikipedia2vec/wikipedia2vec
+
+
+known issues:
+
+ * links inside tags are put into text stream but not link list
+   e.g. ''[[Photometria]]''
+
+
+"""
+import logging
+from typing import Callable, Iterable, List, Optional, Tuple
+import mwparserfromhell
+from mwparserfromhell.nodes import ExternalLink, Heading, Tag, Text, Wikilink
+
+
+logger = logging.getLogger(__name__)
+
+
+FORBIDDEN_SECTIONS = [
+    "bibliography",
+    "citations",
+    "external links",
+    "further reading",
+    "other uses",
+    "references",
+    "see also",
+    "sources",
+]
+
+FORBIDDEN_WIKILINK_PREFIXES = [
+    "file:",
+    "image:",
+    "category:",
+]
+
+ALLOWED_TAGS = [
+    "b",
+    "i",
+    "u",
+]
+
+WikilinkParser = Callable[[Wikilink], Tuple[bool, bool, str, str]]
+
+
+class WikitextPreprocessorMwpfh:
```

Comment:
Seems like we should call this a ParagraphAndLinksParser or something like that.

Reply:
On that note, I think we should rename the WikitextPreprocessor to the QuickAndDirtyTextSimplifier or something like that.  I also think there's some common standards we should try to aim for.  E.g., we could have both of these strategies output a common format that is extended in the case of the ParagraphAndLinksParser that adds fields for links.

Reply:
Yea, I was trying to decide if the underlying parsing tool should be part of the name (mwpfh vs regex) or not.  Couldn't really decide.  Maybe something like `LinkAnnotatedParagraphParser` and `TextStreamParser` (which would make more sense if the tokenization was distinct from the wikitext -> plaintext step)

Reply:
Maybe both could output a list of new line delimitted text chunks (broadly called paragraphs) into something like, 

original regex parser:
```
paragraphs = [
    {"text": "this is text 1"},
    ...]
```
and then the mwpfh version adds the "wikilinks", "section_name", and "section_idx" fields?

Reply:
and then any tokenization step could be applied to the "text" output of either parser